aten/src/ATen/native/cuda/stochastic_rounding.cuh

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <math.h>
+#include <utility>
+
+#include <cuda.h>
+#include <cuda_fp16.h>
+#include <cuda_runtime.h>
+#include <curand.h>
+#include <curand_kernel.h>
+
+#include <ATen/Utils.h>
+#include <ATen/Generator.h>
+#include <ATen/CUDAGeneratorImpl.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAStream.h>
+#include <c10/cuda/CUDAFunctions.h>
+
+// 2^-10 is the step for normal FP16 numbers.
+// 2^-24 is the unit in the last place (ULP)/precision limitation.
+// 24 is **NOT** related to the number of mantissa bits of single precision format.
+__device__ const float TWO_10 = 0.0009765625;
+__device__ const float TWO_24 = 0.000000059604644775390625;
+
+
+template<typename T>
+__device__ __forceinline__ T maybe_upcast(__half x){
+  return T(__half2float(x));
+}
+
+template<>
+__device__ __forceinline__ __half maybe_upcast<__half>(__half x){
+  return x;
+}
+
+__device__ __forceinline__ float get_delta_fp16(float x) {
+  int exponent;
+  frexpf(x, &exponent);
+  exponent -= 1;
+  if (exponent >= -14)
+    return TWO_10 * std::pow(2, exponent);
+  else
+    return TWO_24;
+}
+
+// Natalia magic
+template <typename scalar_t>
+__device__ __forceinline__ scalar_t round_stochastically(float x, float random_value) {
```

Comment:
Right now, this function only rounds to fp16.  I think its syntax is misleading.

To make the usage clearer, and to establish an API that supports stochastic rounding to other types in the future, I think you should define it as follows:
```
template<typename round_to_prec, typename out_type, typename in_type=float>
struct round_stochastically {
  static_assert(false, "round_stochastically only supports round_to_prec=at::Half");
  __device__ __forceinline__ out_type operator()(in_type x, float random_value) {}
};

template <typename out_type, typename in_type=float>
struct round_stochastically<at::Half, out_type, in_type> {
  __device__ __forceinline__ at::Half operator()(in_type x, float random_value) {
    // what we have now
  }
}
```

Then the caller should say
```
weights[i] = round_stochastically<at::Half, scalar_t>(weight, random_values.x);
```

Reply:
L.59' s `maybe_upcast` does cast from `at::Half` to `float`/`double` if necessary and stochastic rounding SGD & Adam kernel use this functionality.

Reply:
We might want to support stochastic rounding to other formats (like bfloat16) later.  The API should allow the caller to select a type that determines the rounding precision, even if the actual rounding code for that precision isn't implemented yet.

Reply:
Understand. Updated.